feat: UIG-2917 - vl-icon - introductie component

ksvlaanderen opened the pull request:
vl-icon-next component
vl-button-next iconen
vl-link-next iconen
generate-icon-mapping.mjs script
Storybook icon-font beheer
Storybook lijst van alle iconen

[Jira](https://www.milieuinfo.be/jira/browse/UIG-2917)
[Bamboo](https://www.milieuinfo.be/bamboo/browse/UIGOV-CUWC362)

Goldflow commented on libs/common/utilities/src/css/vars/typography.var.css.ts:
```diff
@@ -1,8 +1,10 @@
 import { CSSResult, css, unsafeCSS } from 'lit';
-import { fontFamily, serifFontFamily } from '../fonts.css';
+// @ts-ignore: negeer error 'Could not find a declaration file for module' veroorzaakt door het inmporteren van een .js bestand
```
`importeren`

Goldflow commented on libs/components/src/next/icon/stories/vl-icon.stories-arg.ts:
```diff
@@ -0,0 +1,77 @@
+        name: 'icon',
+        description: 'Het icoon dat moet afgebeeld worden.',
+        table: {
+            type: { summary: TYPES.STRING },
```
`type: { summary: TYPES.STRING, required: true},`

Vind het hier wel relevant om dit veld als required in te stellen gezien er anders geen icoon wordt getoond.

Goldflow commented on apps/storybook/docs/c_beheren/5_icon-font.stories.mdx:
```diff
@@ -0,0 +1,39 @@
+
+Er is een centrale plaats in de uigov-web-components repository waar de icon-mapping CSS classes gedefinieerd worden als CSS-in-JS.<br/>
+Elk component dat een icoon nodig heeft importeert dit icon-mapping bestand.<br/>
+Doordat de icon-mapping als CSS-in-JS gedefinieerd is, is er geen duplicatie van CSS code.<br/>
```
Eventueel deze zin uitbreiden met:

Doordat de icon-mapping als CSS-in-JS gedefinieerd is, is er geen duplicatie van CSS code _omdat overal dezelfde instantie van CSS-in-JS object hergebruikt wordt_.

(of iets gelijkaardig)

Goldflow commented on apps/storybook/docs/c_beheren/5_icon-font.stories.mdx:
```diff
@@ -0,0 +1,39 @@
+import {Meta} from '@storybook/addon-docs';
```
💯 Goed gedocumenteerd!

Goldflow commented on libs/integration/src/icon/vl-all-icons.component.ts:
```diff
@@ -0,0 +1,64 @@
+                    <div class="container">
+                        ${allIcons.map((icon) => html`
+                                <div class="icon">
+                                    <vl-icon-next icon=${icon}></vl-icon-next>
```
Zou dat niet cool zijn (en vooral nuttig) als je op een icon klikt, dat je de template direct krijgt in je clipboard?

bv. `<vl-icon-next icon="alarm-silent"></vl-icon-next>`

Dat kan op zich gwn met bv. dit toe te voegen aan click handler:

```ts
 handleClick(icon: string) {
        const clipboardValue = `<vl-icon-next icon="${icon}"></vl-icon-next>`;
        navigator.clipboard.writeText(clipboardValue);
    }
```

Ook mogelijks zeldzaam moment waarbij `cursor:copy` styling relevant zou kunnen zijn ;)

ksvlaanderen replied:
Vind ik een stap te ver gaan, zou leuk zijn maar is waarschijnlijk een hidden feature waar niemand van af weet die enkel zorgt voor extra complexie.

Goldflow replied:
Ja zou dat dan ook wel expliciet vermelden bovenaan de pagina maar 't is sowieso out of scope :)

Ik had de reflex om te denken dat dit er inzat omdat ik dat vaak zie op andere icon pages en dat ik dit intussen verwacht als ik zo'n overzicht zie maar ze kunnen natuurlijk gwn de tekst eronder copypasten.

ksvlaanderen replied:
Voor enkel de naam van het icoon wil ik dat gedrag eventueel wel voorzien.
Dit zou geen regressie kunnen veroorzaken als bv. het vl-icon-next component verandert.

Goldflow commented on libs/integration/src/icon/vl-all-icons.component.ts:
```diff
@@ -0,0 +1,64 @@
+
+        // ! Dit bestand is gegenereerd door het generate-icon-mapping.mjs script
+
```
Prettier formatting wordt niet toegepast op dit document. Vermoed dat dit ook moeilijk te doen valt gezien het bestand gegenereerd wordt door je script?

ksvlaanderen replied:
Valt inderdaad niet te doen tenzij je na het genereren de formatter handmatig laat lopen.

Goldflow commented on libs/common/utilities/src/css/icon/icon.css.ts:
```diff
@@ -0,0 +1,60 @@
+
+        &.vl-icon--light {
+            color: var(--vl-light-text-color);
+        }
```
Graag `vl-icon--light` styling onder `vl-icon--clickable` plaatsen, dan is het mogelijk `light`-styling te combineren met `clickable` styling. Nu is er een css specificity conflict. (Eventueel te bespreken als je die combinatie liever niet als mogelijkheid houdt voor de afnemer.)

![clickable-light](https://github.com/milieuinfo/uigov-web-components/assets/15731348/f4c30995-c1e7-4438-add4-ae55fa433420)

ksvlaanderen replied:
Die combinatie moet niet mogelijk zijn. Een clickable icon heeft een specifieke kleur die we afdwingen.

Goldflow commented on libs/components/src/next/button/stories/vl-button.stories-arg.ts:
```diff
@@ -104,6 +104,34 @@ export const buttonArgTypes: ArgTypes<ButtonArgs> = {
+            defaultValue: { summary: buttonArgs.icon },
+        },
+    },
+    iconAfterText: {
```
Eventueel `iconPlacement` gebruiken ipv `iconAfterText` (met dan bv. `before` als standaard waarde).

Ik geef toe dat dit weinig meerwaarde geeft en dat dit attribuut op zich ook wel duidelijk maakt waarover het gaat maar gezien we `iconPlacement` ook al voor dezelfde functionaliteit gebruiken op andere plaatsen (o.a. `vl-side-sheet`), leek het me mss goed om dit consequent door te trekken.

Eventueel, passen we `iconPlacement` aan (want het is op zich ook raar dat dit geen boolean / flag waarde is)

Maar het is ook ok, om het zo te laten.

Goldflow commented on libs/components/src/next/icon/stories/vl-icon.stories-arg.ts:
```diff
@@ -0,0 +1,77 @@
+            defaultValue: { summary: iconArgs.light },
+        },
+    },
+    before: {
```
Vind `before` & `after` een beetje ongelukkig gekozen attribuut namen. Ik weet dat jij deze niet gekozen hebt en historisch de vorige API volgt maar ik zou eerder opteren voor bv. `left-margin` en `right-margin`.

Behalve mogelijkse verwarring met de pseudo selectors met dezelfde naam (die eerder gaan om de plaatsing van het element zelf) vind ik het een beetje vaag zonder `args` description.

Goldflow commented on libs/components/src/next/link/stories/vl-link.stories-arg.ts:
```diff
@@ -66,6 +66,24 @@ export const linkArgTypes: ArgTypes<LinkArgs> = {
+            defaultValue: { summary: linkArgs.icon },
+        },
+    },
+    iconAfterText: {
```
Als je akkoord gaat met eerdere suggestie ivm `iconPlacement` dan ook eventueel hier toepassen.

Goldflow reviewed (commented):
Uitgetest. Goed werk geleverd!

Technische uitwerking is uitstekend.

Enkele opmerkingen toegevoegd naar ivm API, documentatie en 1 over `light` styling.

Goldflow reviewed (approved):
Sugoi! Bedankt voor al je bloed, zweet en tranen op de `vl-icon` :)

Approved!

ksvlaanderen committed:
feat: UIG-2917 - vl-icon - introductie component

vl-icon-next component
vl-button-next iconen
vl-link-next iconen
generate-icon-mapping.mjs script
Storybook icon-font beheer
Storybook lijst van alle iconen
